src/UIHandler.cpp

```diff
@@ -0,0 +1,345 @@
+{
+    m_thisSize = size;
+
+    if( m_childPanels.empty() ) {
+        return 0;
+    }
+    auto newSize = size;
+
+    if( m_drawBorder )
+        newSize -= { 2, 2 };
+
+    m_childPanels[0]->SetSize( newSize );
+    return 0;
+}
+
+void UIPaddingPanel::DrawEverything( WINDOW *wf_win, point offset )
+{
+
+    if( m_drawBorder ) {
+        UIUtils::DrawBorder( wf_win, offset, m_thisSize );
+    }
+
+    if( !m_childPanels.empty() ) {
+        point offset;
+
+        if( m_drawBorder )
+            offset += { 1, 1 };
+
+        m_childPanels[0]->DrawEverything( wf_win, offset );
+    }
+
+    m_lastSize = m_thisSize;
+}
+
+
+UIPaddingPanel::UIPaddingPanel( bool drawBorder )
+{
+    m_drawBorder = drawBorder;
+}
+
+std::vector<std::shared_ptr<UIPanel>> UITabPanel::GetChild() const
+{
+    return m_childPanels;
+}
+
+void UITabPanel::AddChild( std::shared_ptr<UIPanel> panel )
+{
+    AddChild( std::to_string( m_childPanels.size() ), panel );
+}
+
+void UITabPanel::AddChild( std::string name, std::shared_ptr<UIPanel> panel )
+{
+    m_childPanels.push_back( panel );
+    m_childPanelNames.push_back( name );
+}
+
+void UITabPanel::RemoveChild( size_t index )
+{
+    if( index != m_childPanels.size() - 1 ) {
+        m_childPanels[index] = std::move( m_childPanels.back() );
+    }
+    m_childPanels.pop_back();
+
+    if( index != m_childPanelNames.size() - 1 ) {
+        m_childPanelNames[index] = std::move( m_childPanelNames.back() );
+    }
+    m_childPanelNames.pop_back();
+}
+
+point UITabPanel::RequestedSize( Sizes sizes )
+{
+    point size;
+
+    size += { 0, 2 };
+
+    if( m_drawBorder )
+        size += { 2, 2 };
+
+    if( m_childPanels.empty() ) {
+        return size;
+    }
+
+    auto paneSize = m_childPanels[m_currentTab]->RequestedSize( sizes );
+    size += paneSize;
+
+    int len = 2;
+
+    for( auto tx : m_childPanelNames ) {
+        len += utf8_width( tx ) + 6;
+    }
+
+    if( len > size.x ) {
+        size.x = len;
+    }
+
+    query_yn( std::to_string( size.x ).c_str() );
+    query_yn( std::to_string( size.y ).c_str() );
+
+    return size;
+}
+
+int UITabPanel::SetSize( point size )
+{
+    m_thisSize = size;
+
+    if( m_childPanels.empty() ) {
+        return 0;
+    }
+    auto newSize = size;
+
+    if( m_drawBorder )
+        newSize -= { 2, 2 };
+
+    newSize -= { 0, 2 };
+
+    m_childPanels[m_currentTab]->SetSize( newSize );
+    return 0;
+}
+
+void UITabPanel::DrawEverything( WINDOW *wf_win, point offset )
+{
+    if( m_drawBorder ) {
+        auto bOffset = offset;
+        bOffset += { 0, 2 };
+
+        auto bSize = m_thisSize;
+        bSize -= { 0, 2 };
+
+        UIUtils::DrawBorder( wf_win, bOffset, bSize );
+    }
+
+    int toffset = 2;
+    for( size_t i = 0; i < m_childPanels.size(); i++ ) {
+        toffset += 1;
+
+        UIUtils::DrawTab( wf_win, offset, toffset, ( i == m_currentTab ), m_childPanelNames[i] );
+
+        toffset += 5 + utf8_width( m_childPanelNames[i] );
```

Comment:
Another unexplained magic number. (The number 2 that appears in combination with borders might need explanation as well, through I get what it means.)